src/libtest/formatters.rs

```diff
@@ -0,0 +1,446 @@
+    }
+
+    fn write_test_start(&mut self, _desc: &TestDesc) -> io::Result<()> {
+        // Do not print header, as priting it at this point will result in
+        // an unreadable output when running tests concurrently.
+        Ok(())
+    }
+
+    fn write_result(&mut self, desc: &TestDesc, result: &TestResult, _: &[u8]) -> io::Result<()> {
+        if !(self.terse && desc.name.padding() != PadOnRight) {
+            let name = desc.padded_name(self.max_name_len, desc.name.padding());
+            self.write_plain(&format!("test {} ... ", name))?;
+        }
+
+        match *result {
+            TrOk => self.write_ok(),
+            TrFailed | TrFailedMsg(_) => self.write_failed(),
+            TrIgnored => self.write_ignored(),
+            TrAllowedFail => self.write_allowed_fail(),
+            TrBench(ref bs) => {
+                self.write_bench()?;
+                self.write_plain(&format!(": {}\n", fmt_bench_samples(bs)))
+            }
+        }
+    }
+
+    fn write_timeout(&mut self, desc: &TestDesc) -> io::Result<()> {
+        self.write_plain(&format!("test {} has been running for over {} seconds\n",
+                                  desc.name,
+                                  TEST_WARN_TIMEOUT_S))
+    }
+
+    fn write_run_finish(&mut self, state: &ConsoleTestState) -> io::Result<bool> {
+        if state.options.display_output {
+            self.write_outputs(state)?;
+        }
+        let success = state.failed == 0;
+        if !success {
+            self.write_failures(state)?;
+        }
+
+        self.write_plain("\ntest result: ")?;
+
+        if success {
+            // There's no parallelism at this point so it's safe to use color
+            self.write_pretty("ok", term::color::GREEN)?;
+        } else {
+            self.write_pretty("FAILED", term::color::RED)?;
+        }
+
+        let s = if state.allowed_fail > 0 {
+            format!(
+                ". {} passed; {} failed ({} allowed); {} ignored; {} measured; {} filtered out\n\n",
+                state.passed,
+                state.failed + state.allowed_fail,
+                state.allowed_fail,
+                state.ignored,
+                state.measured,
+                state.filtered_out)
+        } else {
+            format!(
+                ". {} passed; {} failed; {} ignored; {} measured; {} filtered out\n\n",
+                state.passed,
+                state.failed,
+                state.ignored,
+                state.measured,
+                state.filtered_out)
+        };
+
+        self.write_plain(&s)?;
+
+        Ok(success)
+    }
+}
+
+pub(crate) struct JsonFormatter<T> {
+    out: OutputLocation<T>
+}
+
+impl<T: Write> JsonFormatter<T> {
+    pub fn new(out: OutputLocation<T>) -> Self {
+        Self {
+            out,        }
+    }
+
+    fn write_str<S: AsRef<str>>(&mut self, s: S) -> io::Result<()> {
+        self.out.write_all(s.as_ref().as_ref())?;
+        self.out.write_all("\n".as_ref())
+    }
```

Comment:
Some alternative cosmetic suggestions:

* name function `write_message` to hint that a full message, followed by `\n` is written
* accept just `&str` as an argument: I may be mistaking, but seems like we don't really need to be generic here?
* add an `assert!(!s.contains('\n'))`

Reply:
Goo suggestions. Not sure about the usage of the name "message", though. What do we call these json documents we write? I haven't read "message" so far.

_Ideally_, we should call it `write_json_doc` and this method right here serializes a given item to JSON.

Reply:
I think I referred to them as events (mostly).
When serde is eventually used I guess this function will be mostly redundant.

Reply:
>What do we call these json documents we write? 

We call them messages in Cargo:
https://github.com/rust-lang/cargo/blob/2795895b0886e7e4b23577b396a3c1c9c8819abe/src/cargo/util/machine_message.rs#L10 :) 

The problem with `write_str` is that `self.write_str(format(r#"{{ "key": "value" }}"#))` got me thinking that we absolutely have forgotten about the newline, and that's what I want to convey via naming.